# Python Homework 2

##### A stochastic model for a car insurance company's total cost of damages from traffic accidents goes back to the work by Van der Lann and Louter, "A statistical model for the costs of passenger car traffic accidents", Journal of the Royal Statistical Society (1986).

For every $k=1,2,3\ldots$ we denote by the random variable $X_k$ the US dollar amount of a damage from a policy holder's traffic accident which will occur during the year 2022.

We assume that $X_1$, $X_2$,... is an i.i.d. (independent and identically distributed) sequence of exponentially distributed random variables with an average claim size of 1,500 USD. The (random) total number of accidents $N$ in 2022 is assumed to be Poisson distributed with $\lambda$ claims on average.

It is assumed that the number of accidents is independent of the US dollar amount of damages for each accident. That is, the random variable $N$ is independent of the random variables $X_1$, $X_2$,...

The total costs for the insurance company by the end of 2022 will thus be given by the <b>random sum</b> 

$$S_N := X_1 + X_2 + \dots + X_N = \sum_{k = 1}^{N} X_k.$$

Note that the total number $N$ of accidents is random!

The goal of the current exercise is to approximate via simulation:

* the conditional expected total costs $$\mathbb{E}[S_N|N=j]$$ for the insurance company in 2022 given that the number of claims is equal to $j$, and

* the expected total costs $$\mathbb{E}[S_N]$$ for the insurance company in 2022, and

* the probabilities that the total cost will not exceed $K$ USD, i.e., 

$$\mathbb{P}[S_N \leq K] \quad \text{for} \, K = \$30,000,\text{  and  } K=\$45,000$$


As usual, we start with loading some packages:

In [63]:
import numpy as np
import math
np.random.seed(58)

<b>Step 1:</b><br>
First, write a function <tt>randomSum()</tt> which simulates from the distribution of $S_N$ given that $N=j$. The output should just be a single scalar!

<i>Hint:</i> Use proper build-in functions from the <i>NumPy</i>-package in your code in order to sample from an Exponential distribution (check out the GettingStarted python file): 

In [64]:
def randomSum(averageClaimSize, j):
    S_j = np.random.exponential(scale = averageClaimSize, size = j)

    sampleSum = sum(S_j)
        
    return sampleSum  

In [65]:
## Test your function
randomSum(1500,20)

27445.91736556084

<b>Step 2:</b><br>Write a function <tt>simulatorConditional()</tt> which uses the function <tt>randomSum()</tt> to simulate $M \in \mathbb{N}$ samples from the distribution of $S_N$ given that $N=j$. The output should be an array of length $M$.

In [66]:
def simulatorConditional(averageClaimSize, j, M):
    samples = list()

    for i in range (M):
        samples.append(randomSum(averageClaimSize, j))

    samples = np.array(samples)
    
    return samples

In [67]:
## Test your function
simulatorConditional(1500,20,10)

array([17887.06047668, 38309.56588021, 28351.06026457, 26582.43081688,
       41254.65576503, 30219.19712903, 33594.57441019, 23127.93580127,
       22166.24924735, 24552.48280203])

<b>Step 3:</b><br>Use your <tt>simulatorConditional()</tt> function to approximate the conditional expectation $$\mathbb{E}[S_N|N=j]$$ for $j=10$ and $j=20$.

For this, you generate with your function <tt>simulatorConditional()</tt> a bunch of $M=1,000$ independent realizations (samples) $s^{(1)}_j, s^{(2)}_j, \ldots, s^{(M)}_j$ with the distribution of $S_N$ given $N=j$. Then you can approximate:
$$ \$j \cdot 1,500 = \mathbb{E}[S_N|N=j] \approx \frac{1}{M} \sum_{m=1}^M s^{(m)}_j,$$
and check the accuracy of this approximation.

In [68]:
C_E_j_10 = np.mean(simulatorConditional(1500, 10,1000))
C_E_j_20 = np.mean(simulatorConditional(1500, 20,1000))

print("The approximate conditional expectation for j = 10 is {}.".format(C_E_j_10))
print("The approximate conditional expectation for j = 20 is {}.".format(C_E_j_20))

The approximate conditional expectation for j = 10 is 15116.083569199302.
The approximate conditional expectation for j = 20 is 30168.559329139254.


The true conditional expectation for j = 10 is $E[S_N|N=j] = 10 * 1500 = 15000 \approx 15116.083569199302$.
The true conditional expectation for j = 20 is $E[S_N|N=j] = 20 * 1500 = 30000 \approx 30168.559329139254$.

<b>Step 4:</b><br>Write another simulator function which again uses the function <tt>randomSum()</tt> to simulate $M \in \mathbb{N}$ samples now from the unconditional distribution of $S_N$. You will need to write a function <tt>simulator()</tt> that passes Poisson distributed random numbers into the second argument of <tt>randomSum()</tt> now. The output of the function <tt>simulator()</tt> should be an array of length $M$.

In [69]:
def simulator(averageClaimSize, Poissonparameter, M):
    samples = list()

    for i in range(M):
        j_rand = np.random.poisson(lam = Poissonparameter)
        samples.append(randomSum(averageClaimSize, j_rand))

    samples = np.array(samples)
    
    return samples

<b>Step 5:</b><br>We now assume that the average number of claims per year is $20$ and therefore choose $\lambda=20$. As you are seeing in the homework, it holds via the so-called <b>Wald's Identity</b> that the expectation of the random sum $S_N$ is actually given by the formula

\begin{equation}
\mathbb{E}[S_N] = \mathbb{E}[N] \cdot \mathbb{E}[X_1] = 20 \cdot \$1,500 = \$30,000.
\end{equation}

Check via the empirical mean that

$$ \$30,000 = \mathbb{E}[S_N] \approx \frac{1}{M} \sum_{m=1}^M s^{(m)}_N$$

where $s^{(1)}_N, s^{(2)}_N, \ldots, s^{(M)}_N$ denote $M$ independent realizations (samples) from the random variable $S_N$. Use $M = 5000$ simulations.  

In [70]:
E_S_N = np.mean(simulator(1500, 20, 5000))
print("The approximate expectation of the random sum S_N {}.".format(E_S_N))

The approximate expectation of the random sum S_N 30073.123816584313.


<b>Step 6: </b><br>Recall from class that the desired probabilities $\mathbb{P}[S_N \leq K]$ for $K = \$30,000,\text{ and } \$45,000$ can be computed as expectations via an indicator function

$$ \mathbb{P}[S_N \leq K] = \mathbb{E}[1_{\{S_N \leq K\}}]$$

We use once more the empricial mean to approximate

$$ \mathbb{E}[1_{\{S_N \leq K\}}] \approx \frac{1}{M} \sum_{m=1}^M 1_{\{s^{(m)}_N \leq K \}}$$

with $M$ independent realizations (samples) from the random variable $S_N$ (again denoted by $s^{(1)}_N, s^{(2)}_N, \ldots, s^{(M)}_N$).

Write a function <tt>MCprobEstimation()</tt> which estimates the probabilities $\mathbb{P}[S_N \leq K]$ for $K = \$30,000,\text{  and  } K=\$45,000$ as described based on $M$ simulations of $S_N$. The output should be a real number in [0,1]:

In [71]:
def MCprobEstimation(averageClaimSize, averageNumberOfClaims, K, M):
    samples = simulator(averageClaimSize, averageNumberOfClaims, M)

    empiricalSum = 0

    for sample in samples:
        if sample <= K:
            empiricalSum = empiricalSum + 1

    empiricalProb = empiricalSum/M

    return empiricalProb

In [72]:
## Test your function
MCprobEstimation(1500, 20, 20000, 10)

0.2

Test your function with varying $M = 100, 1000, 10000$ simulations:

In [73]:
print(MCprobEstimation(1500, 20, 30000, 100))
print(MCprobEstimation(1500, 20, 30000, 1000))
print(MCprobEstimation(1500, 20, 30000, 10000))

0.5
0.546
0.5311


In [74]:
print(MCprobEstimation(1500, 20, 45000, 100))
print(MCprobEstimation(1500, 20, 45000, 1000))
print(MCprobEstimation(1500, 20, 45000, 10000))

0.89
0.935
0.9325
